### PR TITLE
Misc Additions to Local FPGA Branch

### PR DESCRIPTION
**Pull request by abejgonzalez**
**Related issue**: #666 

**Type of change**: new feature

**Impact**: rtl change + other

**Release Notes**
This builds off of #666 to have it use the Chipyard make system and try to make the compile work with the newest RC changes. Note, this doesn't check for RTL functionality... still needs to be checked on an Arty board.

### CODE REVIEW

File: fpga/.gitignore
Comment (colinschmidt) [inline fpga/.gitignore:1]: I think we should make the builds in here look more standard so we can change this ignore all to something like `generated-src`,`output`, etc.

File: fpga/Makefile
Comment (colinschmidt) [inline fpga/Makefile:51]: Why these included via the standard firrtl method?

File: fpga/Makefile
Comment (colinschmidt) [inline fpga/Makefile:56]: rom->bootrom

File: fpga/Makefile
Comment (colinschmidt) [inline fpga/Makefile:64]: I don't like this name. It is the generated rom not a rom generator. I think `rom_v` `rom_file` etc are better names.

File: fpga/Makefile
Comment (colinschmidt) [inline fpga/Makefile:80]: Does vivado handle duplicates? I think we have had dedupe code for these `*.f` files in the past.

File: fpga/Makefile
Comment (colinschmidt) [inline fpga/Makefile:83]: Again I think we should disambiguate the name of the variable for the output from the name of the target.

File: fpga/Makefile
Comment (colinschmidt) [inline fpga/Makefile:98]: As above. bad name.

File: fpga/Makefile
Comment (colinschmidt) [inline fpga/Makefile:108]: Do we know what this is? Does it apply to the subset we are including?

File: fpga/bootrom/xip/Makefile
Comment (colinschmidt) [inline fpga/bootrom/xip/Makefile:17]: Hopefully we can apply the same `s/filename[^:]/filename_file/` rewrite rule to all these intermediate files.

File: fpga/bootrom/xip/Makefile
Comment (colinschmidt) [inline fpga/bootrom/xip/Makefile:40]: I can't comment on the above lines since they aren't in the diff for this PR, but I have one question.
Is the hardcoded target_address coming from an RTL config fragment? If so we should think about how to do this without repeating ourselves.

File: fpga/bootrom/xip/Makefile
Comment (colinschmidt) [inline fpga/bootrom/xip/Makefile:45]: Nice

File: fpga/src/main/scala/arty/Platform.scala
Comment (colinschmidt) [inline fpga/src/main/scala/arty/Platform.scala:54]: Is this just remnants from some other testing?

File: fpga/src/main/scala/arty/System.scala
Comment (colinschmidt) [inline fpga/src/main/scala/arty/System.scala:38]: It seems weird to support both `bootROM` and `maskROM` above but then not do some connection here.
Do we require a `maskROM` but optionally allow a `bootROM` that never can be the reset location?

File: generators/utilities/src/main/scala/Simulator.scala
Comment (colinschmidt) [inline generators/utilities/src/main/scala/Simulator.scala:89]: Do we need this empty `Seq` at the beginning?

**Review by colinschmidt — APPROVED** [review]
Seem pretty good. I think we don't need to consume the weird style habits in the makefile, and instead can make the names more similar to the rest of chipyard.

Also obviously testing would be good, but I assume that can happen after the merge into the orig PR.

Not blocking merge since this is just a PR to a PR

File: fpga/Makefile
Comment (jerryz123) [inline fpga/Makefile:30]: Should the goal be to eventually eliminate most of these overrides? I believe eventually everything can just live as part of the core chipyard generator.

Comment (colinschmidt) [reply fpga/Makefile:30]: I think thats related to the other checkbox in my review of the original PR. I believe the answer is yes.

File: fpga/.gitignore
Comment (abejgonzalez) [reply fpga/.gitignore:1]: Reasonable goal.

File: fpga/Makefile
Comment (abejgonzalez) [reply fpga/Makefile:30]: Yes. The goal is to get rid of these. I just hard-coded them so that this particular example works all the time. Ideally, this will change with more configuration exposed.

File: fpga/Makefile
Comment (abejgonzalez) [reply fpga/Makefile:51]: These are just black boxes and do not have a resource associated with them. So we would have to make a PR for these to be included through the "normal automatic" way.

Comment (abejgonzalez) [reply fpga/Makefile:51]: For now, I just blindly included them, later I think a PR should be opened in both places to make this automatic.

File: fpga/Makefile
Comment (abejgonzalez) [reply fpga/Makefile:56]: Sure. Just copied the names.

File: fpga/Makefile
Comment (abejgonzalez) [reply fpga/Makefile:64]: Sure. Just copied the names.

File: fpga/Makefile
Comment (abejgonzalez) [reply fpga/Makefile:80]: The `sim_common_files` is deduped. It shouldn't collide with the `all_vsrcs`. As for Vivado... I don't know.

File: fpga/Makefile
Comment (abejgonzalez) [reply fpga/Makefile:108]: I don't know, I just copied this. I think we are just using the Vivado stuff.

File: fpga/bootrom/xip/Makefile
Comment (abejgonzalez) [reply fpga/bootrom/xip/Makefile:40]: Yea. Im not sure. I need to look into this more... they end up using a `MaskROM` over a `BootROM` in the code which is a bit confusing to me.

File: fpga/bootrom/xip/Makefile
Comment (abejgonzalez) [reply fpga/bootrom/xip/Makefile:45]: #vim>emacs

File: fpga/src/main/scala/arty/Platform.scala
Comment (abejgonzalez) [reply fpga/src/main/scala/arty/Platform.scala:54]: Oh. Yea. This should be deleted.

File: fpga/src/main/scala/arty/System.scala
Comment (abejgonzalez) [reply fpga/src/main/scala/arty/System.scala:38]: Yea. This was a hack to try and do what the original code wanted. I need to look into this.

File: generators/utilities/src/main/scala/Simulator.scala
Comment (abejgonzalez) [reply generators/utilities/src/main/scala/Simulator.scala:89]: Sorry? Are you referring to L90? If so, yes, you want to not include the simulation files (or you have to filter them out in the `Makefile` when using `sim_common_files`)

File: fpga/Makefile
Comment (colinschmidt) [reply fpga/Makefile:51]: Issue to track?

File: fpga/Makefile
Comment (colinschmidt) [reply fpga/Makefile:80]: Ok cool as long as we reusing that code.

File: fpga/Makefile
Comment (colinschmidt) [reply fpga/Makefile:108]: I think we should delete it for now, rather than keep around dead code (and perhaps make users think we support this).

File: fpga/bootrom/xip/Makefile
Comment (colinschmidt) [reply fpga/bootrom/xip/Makefile:40]: I believe `MaskROM` implies a specific kind of `ROM` that is reprogrammable with just a single VLSI mask update. I think there is an overload used here sometimes for an FPGA `ROM` that can be updated without the rest of the bit stream being changed.

Besides that definition, I think a `MaskROM` can be a `BootROM` (a `ROM` that contains code to boot, and is where the core resets to), but not all `BootROM`s would be `MaskROM`s and not all `MaskROM`s would be `BootROM`s.

File: generators/utilities/src/main/scala/Simulator.scala
Comment (colinschmidt) [reply generators/utilities/src/main/scala/Simulator.scala:89]: Oh this has changed I think since I looked at it before, or I misread before.
I thought it was:
```
Seq(
) ++ (sim match {
...
```

File: fpga/bootrom/xip/Makefile
Comment (abejgonzalez) [reply fpga/bootrom/xip/Makefile:40]: So it looks like the `MaskROM` is used to load in a ROM after elaboration using `readmemh`. This differs from the BootROM which bakes the `.img` directly into the module in Chisel.

Comment (abejgonzalez) [reply fpga/bootrom/xip/Makefile:40]: Should be good to just swap `MaskROM` for the `BootROM`

**Comment by abejgonzalez** [comment]
Updated. Now uses the CY BootRom, addressed some comments, and now uses `DigitalTop` (aka can now access all CY projects). Need to see what the `MockAON` is since I don't think it can be "disabled" but so far this is getting closer.

**Pull request merged**
